base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java

```diff
@@ -0,0 +1,662 @@
+        order.setStatus(attr.getStringValues().nextElement());
+
+        attr = entry.getAttribute(ATTR_EXPIRES);
+        if (attr != null) {
+            order.setExpirationTime(
+                dateFormat.parse(attr.getStringValues().nextElement()));
+        }
+
+        attr = entry.getAttribute(ATTR_CERTIFICATE_ID);
+        if (attr != null) {
+            order.setCertID(attr.getStringValues().nextElement());
+        }
+
+        List<ACMEIdentifier> identifiers = new ArrayList<>();
+        for (String type : IDENTIFIER_TYPES) {
+            attr = entry.getAttribute(ATTR_IDENTIFIER + ";" + type);
+            for (String value : Collections.list(attr.getStringValues())) {
+                identifiers.add(new ACMEIdentifier(type, value));
+            }
+        }
+        order.setIdentifiers(identifiers.toArray(new ACMEIdentifier[0]));
+
+        attr = entry.getAttribute(ATTR_AUTHORIZATION_ID);
+        order.setAuthzIDs(attr.getStringValueArray());
+
+        return order;
+    }
+
+    public void addOrder(ACMEOrder order) throws Exception {
+        LDAPAttribute[] attrs = {
+            new LDAPAttribute(ATTR_OBJECTCLASS, OBJ_ORDER),
+            new LDAPAttribute(ATTR_ORDER_ID, order.getID()),
+            new LDAPAttribute(ATTR_ACCOUNT_ID, order.getAccountID()),
+            new LDAPAttribute(ATTR_STATUS, order.getStatus()),
+            new LDAPAttribute(ATTR_EXPIRES, dateFormat.format(order.getExpirationTime())),
+            new LDAPAttribute(ATTR_AUTHORIZATION_ID, order.getAuthzIDs())
+        };
+        LDAPAttributeSet attrSet = new LDAPAttributeSet(attrs);
+
+        // identifiers
+        ACMEIdentifier[] identifiers = order.getIdentifiers();
+        for (int i = 0; i < identifiers.length; i++) {
+            attrSet.add(
+                new LDAPAttribute(
+                    ATTR_IDENTIFIER + ";" + identifiers[i].getType(),
+                    identifiers[i].getValue())
+            );
+        }
+
+        String error = order.getError();
+        if (error != null) {
+            attrSet.add(new LDAPAttribute(ATTR_ERROR, error));
+        }
+
+        String dn = ATTR_ORDER_ID + "=" + order.getID() + "," + RDN_ORDER + "," + basedn;
+        LDAPEntry entry = new LDAPEntry(dn, attrSet);
+        ldapAdd(entry);
+    }
+
+    public void updateOrder(ACMEOrder order) throws Exception {
+        String dn = ATTR_ORDER_ID + "=" + order.getID() + "," + RDN_ORDER + "," + basedn;
+        LDAPModificationSet mods = new LDAPModificationSet();
+        mods.add(
+            LDAPModification.REPLACE,
+            new LDAPAttribute(ATTR_STATUS, order.getStatus())
+        );
+        mods.add(
+            LDAPModification.REPLACE,
+            new LDAPAttribute(ATTR_CERTIFICATE_ID, order.getCertID())
+        );
+
+        // update error value
+        String error = order.getError();
+        LDAPAttribute attrError;
+        if (error == null) {
+            attrError = new LDAPAttribute(ATTR_ERROR);  // no values
```

Comment:
Will this remove the error attribute, or just blank the value?

Reply:
I *think* it should remove the attribute.  But I should confirm (and if it does, I will put a comment).

Reply:
Yes, confirmed.  Have elaborated in the comment.